### PR TITLE
feat(HACBS-1010): Create and manage composite snapshots

### PR DESCRIPTION
**Pull request by dirgim**
* Add createCompositeSnapshotsIfConflictExists functionality
* Modify snapshot creation functions, expand labels
* Mark snapshot as failed if superseded by composite snapshot

Signed-off-by: dirgim <kpavic@redhat.com>

### CODE REVIEW

**Review by jinqi7 — APPROVED** [review]
LGTM

File: controllers/pipeline/pipeline_adapter.go [hunk 1]
```diff
@@ -202,6 +226,16 @@ func (a *Adapter) getImagePullSpecFromPipelineRun(pipelineRun *tektonv1beta1.Pip
 	return fmt.Sprintf("%s@%s", strings.Split(outputImage, ":")[0], imageDigest), nil
 }
 
+// getImagePullSpecFromPipelineRun gets the full image pullspec from the given ApplicationSnapshot Component,
```
Comment (hongweiliu17) [inline controllers/pipeline/pipeline_adapter.go:229]:
```suggestion
// getImagePullSpecFromSnapshotComponent gets the full image pullspec from the given ApplicationSnapshot Component,
```

**Review by hongweiliu17 — APPROVED** [review]
LGTM

File: controllers/pipeline/pipeline_adapter.go (hunk 1 above)
Comment (hongweiliu17) [reply controllers/pipeline/pipeline_adapter.go]: This is just a minor issue. But don't why this comment is marked `Resolved` by me when I approve it.

Comment (dirgim) [reply controllers/pipeline/pipeline_adapter.go]: Good catch, will fix that

Comment (dirgim) [reply controllers/pipeline/pipeline_adapter.go]: Fixed the description.

**Review by jsztuka — APPROVED** [review]
lgtm

**Pull request merged**